language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/toml/TomlSnippetManager.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://wso2.com) All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.completions.toml;
+
+import io.ballerina.toml.validator.schema.Schema;
+import org.apache.commons.io.IOUtils;
+import org.ballerinalang.langserver.toml.TomlSyntaxTreeUtil;
+import org.eclipse.lsp4j.CompletionItem;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Maintains all the supported snippets for Toml.
+ *
+ * @since 2.0.0
+ */
+public class TomlSnippetManager {
+
+    private static volatile TomlSnippetManager snippetManager = null;
+    private final Map<Parent, Map<String, CompletionItem>> completions;
+
+    private TomlSnippetManager() {
+        Schema c2cRootSchema = Schema.from(getValidationSchema());
+        TomlSchemaVisitor visitor = new TomlSchemaVisitor();
+        c2cRootSchema.accept(visitor);
+        this.completions = removeRedundantCompletions(visitor.getAllCompletionSnippets());
```

Comment:
Why do we need to remove the redundant completions explicitly? Can't we handle this while computing the completions and adding them?

Reply:
Sorry the naming of the method is misleading. What we do here is, we basically get rid of the tables that only contains another table. That operation should be invoked after adding all the items. Changed the method name.